File: 13week/LeetCode-743-Network-Delay-Time.py
class Solution:
    def __init__(self) -> None:
        self.graph = []
        self.distance = []
        self.visited = []
        self.INF = 999999999
    
    
    def networkDelayTime(self, times: list[list[int]], n: int, k: int) -> int:
        self.graph = [[self.INF] * n for _ in range(n)]
        self.visited = [False] * n

        for u, v, w in times:
            self.graph[u - 1][v - 1] = w

        for i in range(n):
            self.graph[i][i] = 0
        
        def get_small_index():
            min = self.INF
            index = 0
            for i in range(len(self.distance)):
                if self.distance[i] < min and self.visited[i] == False:
                    min = self.distance[i]
                    index = i
            return index
        
        def dijkstra(start):
            self.distance = self.graph[start]
            self.visited[start] = True

            for i in range(n - 2):
                current = get_small_index()
                self.visited[current] = True
                for j in range(n):
                    if self.visited[j] == False:
                        if self.distance[current] + self.graph[current][j] < self.distance[j]:
                            self.distance[j] = self.distance[current] + self.graph[current][j]

            self.visited = [False] * n

        dijkstra(k - 1)
        if self.INF in self.distance:
            return -1
        else:
            return max(self.distance)

File: ShortestPathProblem/Leetcode-743-NetworkDelayTime/Leetcode-743-NetworkDelayTime.cpp
#define pii pair<int, int>

class Solution {
public:
    const int INF = 9999999;
    const static int MAX_N = 101;
    int dist[MAX_N]; // 출발 노드간의 최단 거리를 저장할 배열
    vector<pii> graph[MAX_N];
    
    void initDist() {
        for (int i = 0; i < MAX_N; ++i) dist[i] = INF;
    }
    
    void dijkstra(int start) {  // 출발 노드에서부터 모든 노드까지의 최단 거리 계산
        priority_queue<pii, vector<pii>, greater<pii>> pq;
        pq.emplace(0, start);
        dist[start] = 0;
        
        while (!pq.empty()) {
            int cost = pq.top().first;
            int cur = pq.top().second;
            pq.pop();
           
            if (dist[cur] < cost) continue; // 최단 거리가 구해진 경우 pass
            for (int i = 0; i < graph[cur].size(); ++i) {
                int next = graph[cur][i].first;
                int newCost = cost + graph[cur][i].second;
                if (dist[next] > newCost) {
                    dist[next] = newCost;
                    pq.emplace(newCost, next);
                }
            }
        }
    }

    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        initDist();
        for (int i = 0; i < times.size(); ++i) {
            graph[times[i][0]].push_back({times[i][1], times[i][2]});
        }

        dijkstra(k);
        int answer = 0;
        for (int i = 1; i <= n; ++i) {
            if (dist[i] == INF) return -1; // 도달하지 못하는 노드가 발견된 경우
            answer = max(answer, dist[i]); // 가장 멀리 떨어진 노드 거리 갱신
        }

        return answer;
    }
};